zcash_primitives/src/transaction/builder.rs

```diff
@@ -156,14 +196,52 @@ impl Progress {
     }
 }
 
+/// Rules for how the builder should be configured for each shielded pool.
+#[derive(Clone, Copy)]
+pub enum BuildConfig {
+    Standard {
+        sapling_anchor: sapling::Anchor,
+        orchard_anchor: orchard::Anchor,
+    },
+    Coinbase,
+}
+
+impl BuildConfig {
+    pub fn sapling_builder_config(&self) -> (sapling::builder::BundleType, sapling::Anchor) {
+        match self {
+            BuildConfig::Standard { sapling_anchor, .. } => (
+                sapling::builder::BundleType::DEFAULT,
+                sapling_anchor.clone(),
+            ),
+            BuildConfig::Coinbase => (
+                sapling::builder::BundleType::Coinbase,
+                sapling::Anchor::empty_tree(),
+            ),
+        }
+    }
+
+    pub fn orchard_builder_config(&self) -> (orchard::builder::BundleType, orchard::Anchor) {
+        match self {
+            BuildConfig::Standard { orchard_anchor, .. } => (
+                orchard::builder::BundleType::DEFAULT,
+                orchard_anchor.clone(),
+            ),
+            BuildConfig::Coinbase => (
+                orchard::builder::BundleType::Coinbase,
+                orchard::Anchor::empty_tree(),
+            ),
+        }
+    }
+}
+
 /// Generates a [`Transaction`] from its inputs and outputs.
-pub struct Builder<'a, P, R, U: sapling::builder::ProverProgress> {
+pub struct Builder<'a, P, U: sapling::builder::ProverProgress> {
     params: P,
-    rng: R,
+    build_config: BuildConfig,
     target_height: BlockHeight,
     expiry_height: BlockHeight,
     transparent_builder: TransparentBuilder,
-    sapling_builder: SaplingBuilder,
+    sapling_builder: Option<sapling::builder::Builder>,
     orchard_builder: Option<orchard::builder::Builder>,
```

Comment:
I'm confused now: I thought adding the `bundle_required` boolean meant that we'd be making these builders required (as "optionally produce a bundle" is now supported behaviour by the protocol builders)?

Reply:
So, I debated about how to do this. I decided to leave the Orchard builder as optional, so that it can simply be not present prior to NU5 activation; the alternative is to set it up with a `DISABLED` bundle type. The builder is currently set up that the Sapling builder is always `Some`, but using an Option makes it possible for the Sapling and Orchard code to be structured identically, which is nice for inspection.

If we want to go the other way and use a DISABLED Orchard builder in all cases, I'm happy to do that instead.

Reply:
The impression I got from you yesterday was that making these optional was too complex. If you're happy with this approach, it's fine with me.

Reply:
The difference now is that the anchors are always required at construction; what made things complicated was the caller having control over whether or not to provide a bundle by being able to withhold a choice of anchor. Here the optionality is purely internal.